File: frontend/src/services/reportService.ts
import api from './api'
import type { Report, ReportConfig } from '../types'

export const reportService = {
  getReports: async (): Promise<Report[]> => {
    const response = await api.get('/reports')
    return response.data.data
  },

  getReport: async (reportId: number): Promise<Report> => {
    const response = await api.get(`/reports/${reportId}`)
    return response.data.data
  },

  createReport: async (data: {
    name: string
    description: string
    config: ReportConfig
    created_by?: string
  }): Promise<Report> => {
    const response = await api.post('/reports', data)
    return response.data.data
  },

  updateReport: async (
    reportId: number,
    data: {
      name?: string
      description?: string
      config?: ReportConfig
    }
  ): Promise<Report> => {
    const response = await api.put(`/reports/${reportId}`, data)
    return response.data.data
  },

  deleteReport: async (reportId: number): Promise<void> => {
    await api.delete(`/reports/${reportId}`)
  },
}

File: app/services/report_service.py
from app.models.report import Report

class ReportService:
    @staticmethod
    def create_report(name: str, description: str, config: dict, created_by: str = '') -> Report:
        report = Report(
            name=name,
            description=description,
            config=config,
            created_by=created_by,
        )
        return report.save()
    
    @staticmethod
    def get_report(report_id: int) -> Report:
        return Report.get_by_id(report_id)
    
    @staticmethod
    def get_all_reports() -> list:
        return Report.get_all()
    
    @staticmethod
    def update_report(report_id: int, name: str = None, description: str = None, config: dict = None) -> Report:
        report = Report.get_by_id(report_id)
        if not report:
            raise ValueError(f"Report {report_id} not found")
        
        if name is not None:
            report.name = name
        if description is not None:
            report.description = description
        if config is not None:
            report.config = config
        
        return report.save()
    
    @staticmethod
    def delete_report(report_id: int) -> bool:
        report = Report.get_by_id(report_id)
        if not report:
            return False
        return report.delete()

